h/api/storage.py

```diff
@@ -114,42 +118,60 @@ def create_annotation(request, data):
-        request.db,
-        annotation.target_uri,
-        [u['uri'] for u in document_uri_dicts],
-        created=annotation.created,
-        updated=annotation.updated)
-
-    if documents.count() > 1:
-        document = models.merge_documents(request.db,
-                                          documents,
-                                          updated=annotation.updated)
-    else:
-        document = documents.first()
-
-    document.updated = annotation.updated
-
-    for document_uri_dict in document_uri_dicts:
-        models.create_or_update_document_uri(
-            session=request.db,
-            document=document,
-            created=annotation.created,
-            updated=annotation.updated,
-            **document_uri_dict)
-
-    for document_meta_dict in document_meta_dicts:
-        models.create_or_update_document_meta(
-            session=request.db,
-            document=document,
-            created=annotation.created,
-            updated=annotation.updated,
-            **document_meta_dict)
+    models.update_document_metadata(
+        request.db, annotation, document_meta_dicts, document_uri_dicts)
+
+    return annotation
+
+
+def update_annotation(session, annotation_id, data):
+    """
+    Update an existing annotation and its associated document metadata.
+
+    Update the annotation identified by annotation_id with the given
+    data. Create, delete and update document metadata as appropriate.
+
+    :param session: the database session
+    :type session: sqlalchemy.orm.session.Session
+
+    :param annotation_id: the ID of the annotation to be updated, this is
+        assumed to be a validated ID of an annotation that does already exist
+        in the database
+    :type annotation_id: string
+
+    :param data: the validated data with which to update the annotation
+    :type data: dict
+
+    :returns: the updated annotation
+    :rtype: h.api.models.Annotation
+
+    """
+    # Remove any 'document' field first so that we don't try to save it on the
+    # annotation object.
+    document = data.pop('document', None)
+
+    annotation = models.Annotation.query.get(annotation_id)
+
+    # Just modifying the annotation.extra dict in place doesn't let sqlalchemy
+    # know that it has changed and needs to be saved to the database.
+    # Work around this for now by copying, modifying then assigning.
+    extra = dict(annotation.extra)
+    extra.update(data.pop('extra', {}))
+    annotation.extra = extra
+
+    for key, value in data.items():
+        setattr(annotation, key, value)
+
+    if document:
+        document_uri_dicts = document['document_uri_dicts']
+        document_meta_dicts = document['document_meta_dicts']
+        models.update_document_metadata(
+            session, annotation, document_meta_dicts, document_uri_dicts)
 
     return annotation
 
 
-def update_annotation(request, id, data):
+def legacy_update_annotation(request, annotation_id, data):
```

Comment:
Same question as above. I'm not sure `annotation_id` improves the readability of this at all... to me it's just "stuttering" like `conn_t conn = (conn_t) { ... }` in C.